collect_app/src/test/java/org/odk/collect/android/formentry/QuitFormDialogFragmentTest.java

```diff
@@ -0,0 +1,92 @@
+import androidx.fragment.app.FragmentManager;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.odk.collect.android.R;
+import org.odk.collect.android.formentry.saving.FormSaveViewModel;
+import org.odk.collect.android.support.RobolectricHelpers;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.shadows.ShadowDialog;
+
+import static android.view.View.GONE;
+import static android.view.View.VISIBLE;
+import static junit.framework.TestCase.assertTrue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.odk.collect.android.support.RobolectricHelpers.mockViewModelProvider;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricTestRunner.class)
+public class QuitFormDialogFragmentTest {
+
+    private FragmentActivity activity;
+    private FragmentManager fragmentManager;
+    private QuitFormDialogFragment dialogFragment;
+
+    private FormSaveViewModel viewModel;
+
+    @Before
+    public void setup() {
+        activity = RobolectricHelpers.createThemedActivity(FragmentActivity.class);
+        fragmentManager = activity.getSupportFragmentManager();
+        dialogFragment = new QuitFormDialogFragment();
+
+        viewModel = mockViewModelProvider(activity, FormSaveViewModel.class).get(FormSaveViewModel.class);
+    }
+
+    @Test
+    public void shouldShowCorrectView() {
+        dialogFragment.show(fragmentManager, "TAG");
+        AlertDialog dialog = (AlertDialog) dialogFragment.getDialog();
+
+        assertNotNull(dialog);
+        assertThat(dialog.getButton(DialogInterface.BUTTON_NEGATIVE).getVisibility(), equalTo(GONE));
+        assertThat(dialog.getButton(DialogInterface.BUTTON_POSITIVE).getVisibility(), equalTo(VISIBLE));
+        assertThat(dialog.getButton(DialogInterface.BUTTON_POSITIVE).getText(),
+                equalTo(activity.getString(R.string.do_not_exit)));
+
+        String title = dialogFragment.getTitle(viewModel);
+        dialogFragment.setTitle();
+        assertThat(shadowOf(dialogFragment.getDialog()).getTitle(), equalTo(title));
+    }
+
+    @Test
+    public void dialogIsCancellable() {
+        dialogFragment.show(fragmentManager, "tag");
+        assertThat(shadowOf(dialogFragment.getDialog()).isCancelable(), equalTo(true));
+    }
+
+    @Test
+    public void clickingCancel_shouldDismissTheDialog() {
+        dialogFragment.show(fragmentManager, "tag");
+        AlertDialog dialog = (AlertDialog) ShadowDialog.getLatestDialog();
+        assertTrue(dialog.isShowing());
+
+        dialog.getButton(DialogInterface.BUTTON_POSITIVE).performClick();
+        assertFalse(dialog.isShowing());
+        assertTrue(shadowOf(dialog).hasBeenDismissed());
+    }
+
+    @Test
+    public void restoringFragment_retainsTitle() {
```

Comment:
Given that the only "state" in this fragment comes from the ViewModel does it make sense to have this test? There isn't any mutable state that needs to be carried from one instance of the fragment to the other as far as I can see.

Reply:
Yes, right @seadowg. Checking it just once would be fine.

Reply:
Also I think the part that we are checking that whether the dialog fragment shows the correct title or not can be covered with the test of the ViewModel (whether it returns the correct form name)?

Reply:
Yeah as above we just need that test I think!

Reply:
Then, should we even remove the test to check the title of the dialog once, as it can be covered in the tests for the ViewModel?

Reply:
Yeah I think that make sense

Reply:
Ach I'm so sorry I misinterpreted your question! My bad for skimming. 

Try removing line 60 from `QuitFormDialogFragment` and then running the tests to see if you have coverage of the title being displayed in the dialog. I don't think you do now you've removed that test. The ViewModel tests check that the `ViewModel.getFormLabel()` does the right thing but here you're testing the dialog, not the view model.

Reply:
Yes sorry @seadowg! I was a little confused too.